File: perl/queens.pl
# Copyright 2018 John Hurst
# John Hurst
# 2018-12-22

use strict;
use warnings;
use Time::HiRes qw(time);

sub ok {
  my ($size, $row, $cols, $diags1, $diags2, $col) = @_;
  return ($cols & (1 << $col)) == 0 &&
         ($diags1 & (1 << $row + $col)) == 0 &&
         ($diags2 & (1 << $row - $col + $size - 1)) == 0;
}

sub place {
  my ($size, $row, $cols, $diags1, $diags2, $col) = @_;
  return (
    $size,
    $row + 1,
    $cols | (1 << $col),
    $diags1 | (1 << $row + $col),
    $diags2 | (1 << $row - $col + $size - 1)
  );
}

sub solve {
  my ($size, $row, $cols, $diags1, $diags2) = @_;
  if ($row == $size) {
    return 1;
  }
  my $result = 0;
  foreach my $col (1..$size) {
    if (ok($size, $row, $cols, $diags1, $diags2, $col)) {
      $result += solve(place($size, $row, $cols, $diags1, $diags2, $col));
    }
  }
  return $result;
}

my $from = $#ARGV >= 0 ? $ARGV[0] : 8;
my $to = $#ARGV >= 1 ? $ARGV[1] : $from;

foreach my $size ($from..$to) {
  my $start = time;
  my $result = solve($size, 0, 0, 0, 0);
  my $duration = time - $start;
  printf("%d,%d,%0.3f\n", $size, $result, $duration);
}

File: elixir/lib/Queens.ex
# Copyright 2018 John Hurst
# JOhn Hurst (john.b.hurst@gmail.com)
# 2018-12-22

defmodule Queens do
  use Bitwise

  def ok(board, col) do
    (board.cols &&& (1 <<< col)) == 0 and
    (board.diags1 &&& (1 <<< board.row + col)) == 0 and
    (board.diags2 &&& (1 <<< board.row - col + board.size - 1)) == 0
  end

  def place(board, col) do
    %Board{
      size: board.size,
      row: board.row + 1,
      cols: board.cols ||| (1 <<< col),
      diags1: board.diags1 ||| (1 <<< board.row + col),
      diags2: board.diags2 ||| (1 <<< board.row - col + board.size - 1)
    }
  end

  def solve(board) do
    if board.row == board.size, do: 1, else: 1..board.size
      |> Enum.filter(fn col -> ok(board, col) end)
      |> Enum.map(fn col -> solve(place(board, col)) end)
      |> Enum.sum
  end

  def main(argv) do
    from = if length(argv) > 0, do: String.to_integer(Enum.at(argv, 0)), else: 8
    to = if length(argv) > 1, do: String.to_integer(Enum.at(argv, 1)), else: from
    for size <- from..to do
      start = Time.utc_now
      result = solve(%Board{size: size})
      duration = Time.diff(Time.utc_now, start, 1000)/1000
      IO.puts "#{size},#{result},#{duration}"
    end
  end
end
